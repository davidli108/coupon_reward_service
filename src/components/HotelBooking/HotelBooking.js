// @flow
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import { withTranslation } from 'react-i18next';
import styled, { keyframes } from 'styled-components';
import breakpoint from 'styled-components-breakpoint';
import { FiSearch } from 'react-icons/fi';
import axios from 'axios';
import { getIsAuthenticated } from '@modules/auth/AuthReducer';
import AppConfig from '@config/AppConfig';

type HotelBookingProps = {
  isAuthenticated: boolean,
  toggleLogin: Function,
};

const HotelBooking = ({ isAuthenticated, toggleLogin }: HotelBookingProps) => {
  const [isActive, setIsActive] = useState(false);
  const [code, setCode] = useState('');
  const [hasError, setHasError] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  const toggleActive = () => {
    setIsActive(!isActive);
  };

  const submitForm = (e: Event) => {
    e.preventDefault();
    setIsLoading(true);
    axios.post(`${AppConfig.apiUrl}/bookings/finds`, { code }).then(
      (response: any) => {
        if (response.success && response.data.url !== '') {
          setHasError(false);
          window.top.location.href = response.data.url;
        }
        setIsLoading(false);
      },
      () => {
        setHasError(true);
        setIsLoading(false);
      },
    );
  };

  const handleChange = e => {
    setCode(e.target.value);
  };

  return (
    <HotelBooking.Wrapper>
      <HotelBooking.Button isActive={isActive} onClick={toggleActive}>
        Hotel Booking
      </HotelBooking.Button>
      <HotelBooking.Popup isActive={isActive}>
        {isAuthenticated ? (
          <p>
            To manage your hotel booking please enter your confirmation number
            below.
          </p>
        ) : (
          <p>
            To manage your hotel booking,{' '}
            <span onClick={toggleLogin}>login to your account</span> or enter
            your confirmation number below.
          </p>
        )}
        {hasError && <b>Please enter a valid confirmation number.</b>}
        <HotelBooking.Form onSubmit={submitForm} isLoading={isLoading}>
          <input
            type="text"
            maxLength={20}
            value={code}
            onChange={handleChange}
            placeholder="Enter confirmation #"
          />
          <HotelBooking.FormSubmit>
            {isLoading ? <HotelBooking.Loading /> : <FiSearch />}
          </HotelBooking.FormSubmit>
        </HotelBooking.Form>
      </HotelBooking.Popup>
      <HotelBooking.Backdrop onClick={toggleActive} isActive={isActive} />
    </HotelBooking.Wrapper>
  );
};

HotelBooking.Wrapper = styled.div`
  position: relative;

  ${breakpoint('xs')`
    height: auto !important;
    min-width: 290px;
  `}

  ${breakpoint('lg')`
    height: 100% !important;
    min-width: 0;
  `}
`;

HotelBooking.Button = styled.button`
  display: flex;
  align-items: center;
  border: 0;
  background-color: transparent;

  width: 100%;
  height: 100%;
  padding: 0 23px;
  letter-spacing: 0.6px;
  font-weight: 500;
  line-height: 25px;

  color: white;
  font-family: Roboto, Arial, sans-serif;

  cursor: pointer;

  &:focus {
    outline: none;
  }

  ${breakpoint('xs')`
    margin: 20px 0 0;
    font-weight: 700;
    font-size: 16px;
    line-height: 19px;
  `}

  ${breakpoint('lg')`
    margin: 0;
    font-size: 19.5px;
    background-color: ${({ isActive }) =>
      isActive ? '#03b6d1' : 'transparent'};

    &:hover {
      background: #02a6bf;
    }
  `}
`;

HotelBooking.Popup = styled.div`
  box-sizing: border-box;

  p {
    color: #fff;
    letter-spacing: 0.5px;
    font-weight: 500;
    margin: 10px 0 0;

    ${breakpoint('xs')`
      font-size: 12px;
      line-height: 17px;
      display: block;
    `}

    ${breakpoint('lg')`
      text-align: center;
      color: #62707B;
      padding: 0 16px;
      margin: 0;
    `}

    span {
      cursor: pointer;
      text-decoration: underline;
      font-weight: 700;
    }
  }

  b {
    display: block;
    font-size: 12px;
    line-height: 17px;
    letter-spacing: 0.5px;
    margin-top: 10px;
    font-weight: 500;
    color: #a94442;

    ${breakpoint('lg')`
      color: red;
      font-weight: 700;
    `}
  }

  ${breakpoint('xs')`
    display: block;
    position: static;
    padding: 0 23px;
  `}

  ${breakpoint('lg')`
    display: ${({ isActive }) => (isActive ? 'block' : 'none')};
    position: absolute;
    width: 366px;
    border-radius: 5px;
    -webkit-box-shadow: 0 7px 14.4px 1.6px rgba(0,0,0,.19);
    top: 65px;
    position: absolute;
    background: #fff;
    z-index: 10;
    left: 50%;
    transform: translateX(-50%);
    margin-top: 20px;
    padding: 20px 17px 12px;

    &::after {
      content: '';
      display: block;
      position: absolute;
      left: 50%;
      bottom: 100%;
      transform: translateX(-50%);
      border: 9px solid transparent;
      border-bottom-color: #fff;
    }
  `}
`;

HotelBooking.Form = styled.form`
  display: flex;
  background: #fbfbfb;
  border: 1px solid #59c6e1;
  box-sizing: border-box;
  border-radius: 5px;
  margin: 0;
  font-weight: 400;
  font-size: 14px;
  line-height: 16px;
  letter-spacing: 0.5px;
  color: #899197;
  padding: 0;
  position: relative;
  height: 41px;
  pointer-events: ${({ isLoading }) => (isLoading ? 'none' : 'auto')};

  input {
    background-color: transparent;
    border: 0;
    flex: 1;
    height: 100%;
    padding: 11px 13px 12px 15px;
    box-sizing: border-box;
    font-weight: 400;
    font-size: 12px;
    line-height: 16px;
    letter-spacing: 0.5px;
    color: #899197;

    &:focus {
      outline: none;
    }

    &::placeholder {
      color: #ccc;
    }

    ${breakpoint('lg')`
      font-size: 14px;
    `}
  }

  ${breakpoint('lg')`
    margin: 13px 0 0;
  `}
`;

HotelBooking.FormSubmit = styled.button`
  width: 39px;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 0;
  background-color: transparent;
  cursor: pointer;

  svg {
    width: 16px;
    height: 16px;
    color: #ccc;
  }

  &:hover {
    svg {
      color: ${props => props.theme.colors.blue};
    }
  }

  &:focus {
    outline: none;
  }
`;

HotelBooking.Backdrop = styled.div`
  pointer-events: ${({ isActive }) => (isActive ? 'auto' : 'none')};
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 9;
`;

const spin = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

HotelBooking.Loading = styled.div`
  width: 20px;
  height: 20px;
  border: 2px solid #ccc;
  border-left-color: ${props => props.theme.colors.blue};
  border-top-color: ${props => props.theme.colors.blue};
  border-radius: 50%;
  animation: ${spin} 0.4s linear infinite;
`;

const mapStateToProps = state => ({
  isAuthenticated: getIsAuthenticated(state),
});
export default compose(
  connect(
    mapStateToProps,
    null,
  ),
  withTranslation(),
)(HotelBooking);
