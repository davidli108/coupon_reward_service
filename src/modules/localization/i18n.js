import i18n from 'i18next';
import parseDomain from 'parse-domain';
import * as R from 'ramda';
import { initReactI18next } from 'react-i18next';

import translationDE from './locales/de/translation';
import translationEN from './locales/en/translation';
import translationFR from './locales/fr/translation';
import translationGB from './locales/gb/translation';

const resources = {
  de: {
    translation: translationDE,
  },
  en: {
    translation: translationEN,
  },
  fr: {
    translation: translationFR,
  },
  gb: {
    translation: translationGB,
  },
};

const Locale = {
  en: 'en',
  de: 'de',
  fr: 'fr',
  gb: 'gb',
};

const localeToDomain = {
  com: Locale.en,
  de: Locale.de,
  fr: Locale.fr,
  'co.uk': Locale.gb,
};

const fallbackLocale = Locale.en;

export const getDomainAttrs = () => {
  const domain = window.location.hostname;
  const protocol = window.location.protocol;

  const parsedDomain = parseDomain(domain);
  const tld = R.propOr(null, 'tld', parsedDomain);
  const enDomain = domain.replace(tld, 'com');
  const enOrigin = `${protocol}//${enDomain}`;

  return {
    domain,
    protocol,
    tld,
    enDomain,
    enOrigin,
  };
};

export const redirectToEnOrigin = () => {
  const domainAttrs = getDomainAttrs();
  window.location.href = domainAttrs.enOrigin;
};

export const getLocale = () => {
  const domainAttrs = getDomainAttrs();
  return localeToDomain[domainAttrs.tld] || fallbackLocale;
};

export const getLocaleConfig = () => {
  return {
    // TODO: enable authentication on all domains after feature support added
    isAuthenticationAvailable: true,
    // TODO: enable follow store on all domains after feature support added
    isFollowStoreAvailable: true,
  };
};

export const initializeI18n = () => {
  // eslint-disable-next-line import/no-named-as-default-member
  i18n.use(initReactI18next).init({
    resources,
    lng: getLocale(),
    fallbackLng: fallbackLocale,
    keySeparator: '.',
    interpolation: {
      escapeValue: false,
    },
  });
};

export default i18n;
