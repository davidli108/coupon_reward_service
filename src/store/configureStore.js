// @flow
import { routerMiddleware } from 'connected-react-router';
import { createBrowserHistory, createMemoryHistory } from 'history';
import { applyMiddleware, compose, createStore } from 'redux';
import { persistStore } from 'redux-persist';
import createSagaMiddleware from 'redux-saga';
import thunk from 'redux-thunk';

import axiosMiddleware from './middlewares/axiosMiddleware';
import loggerMiddleware from './middlewares/loggerMiddleware';
import promiseMiddleware from './middlewares/promiseMiddleware';
import rootReducer from './rootReducer';
import rootSaga from './rootSaga';

export const isServer = !(
  typeof window !== 'undefined' &&
  window.document &&
  window.document.createElement
);

export default () => {
  // const history = createHistory();
  const history = isServer
    ? createMemoryHistory({ initialEntries: ['/'] })
    : createBrowserHistory();

  const initialState = !isServer ? window.__PRELOADED_STATE__ || {} : {};

  const sagaMiddleware = createSagaMiddleware();

  const middlewares = [
    routerMiddleware(history),
    axiosMiddleware,
    thunk,
    promiseMiddleware,
    sagaMiddleware,
  ];

  if (!isServer && typeof jest === 'undefined') {
    middlewares.push(loggerMiddleware);
  }

  const composeEnhancers = isServer
    ? compose
    : window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  // $FlowFixMe
  const store = createStore(
    (rootReducer: Object)(history),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares)),
  );
  const persistor = persistStore(store);

  sagaMiddleware.run(rootSaga);

  // $FlowFixMe
  if (module.hot) {
    // $FlowFixMe
    module.hot.accept('./rootReducer', () => {
      // eslint-disable-next-line global-require
      const nextRootReducer: Object = require('./rootReducer').default(history);
      store.replaceReducer(nextRootReducer);
    });
  }

  return { history, persistor, store };
};
