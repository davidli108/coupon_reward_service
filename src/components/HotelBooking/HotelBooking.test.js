import React from 'react';
import HotelBooking from './HotelBooking';
import renderer from 'react-test-renderer';
import { initializeI18n } from '../../modules/localization/i18n';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { getTheme } from '@theme';
import { Provider } from 'react-redux';
import configureStore from '../../store/configureStore';
const { theme } = getTheme('base');

initializeI18n();
const { store } = configureStore();

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HotelBooking isAuthenticated={true} toggleLogin={() => {}} />
        </ThemeProvider>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
