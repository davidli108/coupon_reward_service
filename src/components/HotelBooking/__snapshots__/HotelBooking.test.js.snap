// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  position: relative;
  height: auto !important;
  min-width: 290px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  background-color: transparent;
  width: 100%;
  height: 100%;
  padding: 0 23px;
  -webkit-letter-spacing: 0.6px;
  -moz-letter-spacing: 0.6px;
  -ms-letter-spacing: 0.6px;
  letter-spacing: 0.6px;
  font-weight: 500;
  line-height: 25px;
  color: white;
  font-family: Roboto,Arial,sans-serif;
  cursor: pointer;
  margin: 20px 0 0;
  font-weight: 700;
  font-size: 16px;
  line-height: 19px;
}

.c1:focus {
  outline: none;
}

.c2 {
  box-sizing: border-box;
  display: block;
  position: static;
  padding: 0 23px;
}

.c2 p {
  color: #fff;
  -webkit-letter-spacing: .5px;
  -moz-letter-spacing: .5px;
  -ms-letter-spacing: .5px;
  letter-spacing: .5px;
  font-weight: 500;
  margin: 10px 0 0;
  font-size: 12px;
  line-height: 17px;
  display: block;
}

.c2 p span {
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-weight: 700;
}

.c2 b {
  display: block;
  font-size: 12px;
  line-height: 17px;
  -webkit-letter-spacing: .5px;
  -moz-letter-spacing: .5px;
  -ms-letter-spacing: .5px;
  letter-spacing: .5px;
  margin-top: 10px;
  font-weight: 500;
  color: #a94442;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #FBFBFB;
  border: 1px solid #59C6E1;
  box-sizing: border-box;
  border-radius: 5px;
  margin: 0;
  font-weight: 400;
  font-size: 14px;
  line-height: 16px;
  -webkit-letter-spacing: .5px;
  -moz-letter-spacing: .5px;
  -ms-letter-spacing: .5px;
  letter-spacing: .5px;
  color: #899197;
  padding: 0;
  position: relative;
  box-sizing: border-box;
  height: 41px;
  pointer-events: auto;
}

.c3 input {
  background-color: transparent;
  border: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  padding: 11px 13px 12px 15px;
  box-sizing: border-box;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: .5px;
  -moz-letter-spacing: .5px;
  -ms-letter-spacing: .5px;
  letter-spacing: .5px;
  color: #899197;
}

.c3 input:focus {
  outline: none;
}

.c3 input::-webkit-input-placeholder {
  color: #ccc;
}

.c3 input::-moz-placeholder {
  color: #ccc;
}

.c3 input:-ms-input-placeholder {
  color: #ccc;
}

.c3 input::placeholder {
  color: #ccc;
}

.c4 {
  width: 39px;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  background-color: transparent;
  cursor: pointer;
}

.c4 svg {
  width: 16px;
  height: 16px;
  color: #ccc;
}

.c4:hover svg {
  color: #00CBE9;
}

.c4:focus {
  outline: none;
}

.c5 {
  pointer-events: none;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 9;
}

@media (min-width:62em) {
  .c0 {
    height: 100% !important;
    min-width: 0;
  }
}

@media (min-width:62em) {
  .c1 {
    margin: 0;
    font-size: 19.5px;
    background-color: transparent;
  }

  .c1:hover {
    background: #02a6bf;
  }
}

@media (min-width:62em) {
  .c2 p {
    text-align: center;
    color: #62707B;
    padding: 0 16px;
    margin: 0;
  }
}

@media (min-width:62em) {
  .c2 b {
    color: red;
    font-weight: 700;
  }
}

@media (min-width:62em) {
  .c2 {
    display: none;
    position: absolute;
    width: 366px;
    border-radius: 5px;
    -webkit-box-shadow: 0 7px 14.4px 1.6px rgba(0,0,0,.19);
    top: 65px;
    position: absolute;
    background: #fff;
    z-index: 10;
    left: 50%;
    -webkit-transform: translateX(-50%);
    -ms-transform: translateX(-50%);
    transform: translateX(-50%);
    margin-top: 20px;
    padding: 20px 17px 12px;
  }

  .c2::after {
    content: '';
    display: block;
    position: absolute;
    left: 50%;
    bottom: 100%;
    -webkit-transform: translateX(-50%);
    -ms-transform: translateX(-50%);
    transform: translateX(-50%);
    border: 9px solid transparent;
    border-bottom-color: #fff;
  }
}

@media (min-width:62em) {
  .c3 input {
    font-size: 14px;
  }
}

@media (min-width:62em) {
  .c3 {
    margin: 13px 0 0;
  }
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    Hotel Booking
  </button>
  <div
    className="c2"
  >
    <p>
      To manage your hotel booking, 
      <span
        onClick={[Function]}
      >
        login to your account
      </span>
       or enter your confirmation number below.
    </p>
    <form
      className="c3"
      onSubmit={[Function]}
    >
      <input
        maxLength={20}
        onChange={[Function]}
        placeholder="Enter confirmation #"
        type="text"
        value=""
      />
      <button
        className="c4"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
      </button>
    </form>
  </div>
  <div
    className="c5"
    onClick={[Function]}
  />
</div>
`;
